packages/server-reason-react-ppx/server_reason_react_ppx.ml

```diff
@@ -854,7 +856,213 @@ let rewrite_jsx =
+        let action = ReactServerDOMEsbuild.createServerReference [%e estring ~loc id] in
+        ([%e pexp_apply ~loc [%expr action] apply_args] [@u])]
+    in
+    fn
+
+  let transform_client_function nested_modules structure_item =
+    match structure_item.pstr_desc with
+    | Pstr_value (rec_flag, value_bindings) when isReactServerFunctionBinding (List.hd value_bindings) ->
+        let loc = structure_item.pstr_loc in
+        let vb = List.hd value_bindings in
+
+        let function_name = get_function_name vb in
+        let args = get_arguments vb.pvb_expr |> List.map ~f:get_arg_details |> List.rev in
+        let base_fn = vb.pvb_expr in
+        let id = generate_id ~loc:vb.pvb_loc function_name in
+        let value_binding =
+          {
+            vb with
+            pvb_expr =
+              create_server_function_contract ~loc:vb.pvb_loc id
+                [%expr [%e last_expr_to_fn ~loc base_fn (create_client_function ~loc id args)]];
+          }
+        in
+
+        let loc = structure_item.pstr_loc in
+        let module_name = String.concat "." nested_modules in
+        let comment = Printf.sprintf "// extract-server-function %s %s %s" id function_name module_name in
+        let raw = estring ~loc comment in
+        let extract_client_raw = [%stri [%%raw [%e raw]]] in
+        [%stri
+          include struct
+            [%%i extract_client_raw]
+            [%%i pstr_value ~loc:structure_item.pstr_loc rec_flag [ value_binding ]]
+          end]
+    | _ -> structure_item
+
+  let traverse =
+    object (_)
+      inherit [Expansion_context.Base.t] Ppxlib.Ast_traverse.map_with_context as super
+      val mutable nested_modules = []
+
+      method! structure_item ctx structure_item =
+        let structure_item = super#structure_item ctx structure_item in
+        match !mode with
+        | Native -> transform_native_function structure_item
+        | Js -> transform_client_function nested_modules structure_item
+
+      method! module_binding ctx module_binding =
+        match !mode with
+        | Native -> super#module_binding ctx module_binding
+        | Js -> (
+            match module_binding.pmb_name.txt with
+            | Some name ->
+                nested_modules <- nested_modules @ [ name ];
+                let module_binding = super#module_binding ctx module_binding in
+                nested_modules <- List.tl nested_modules;
+                module_binding
+            | _ -> super#module_binding ctx module_binding)
+    end
+end
+
 let () =
   Driver.add_arg "-melange" (Unit (fun () -> mode := Js)) ~doc:"preprocess for js build";
-  Ppxlib.Driver.V2.register_transformation "server-reason-react.ppx" ~preprocess_impl:rewrite_jsx#structure
+  Ppxlib.Driver.V2.register_transformation "server-reason-react.ppx"
+    ~preprocess_impl:(fun ctx str ->
+      let str = rewrite_jsx#structure ctx str in
+      let str = ServerFunction.traverse#structure ctx str in
```

Comment:
why did you choose to create a new traverse, instead of adding everything in rewrite_jsx? (and rename it?)

I remember you mention it in a meeting, but now we are doing the double of iterations over an AST.